frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -0,0 +1,103 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import com.google.common.collect.Tables;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.lang.RuntimeException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit)
+      throws PlaceVisitAlreadyExistsException, Throwable {
+    try {
+      storage.compute(placeVisit.tripId(), (tripKey, placesMap) -> {
+        if (placesMap == null) {
+          Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+          newPlaceMap.put(placeVisit.placeId(), placeVisit);
+          return newPlaceMap;
+        }
+        if (placesMap.get(placeVisit.placeId()) != null) {
+          throw new RuntimeException(new PlaceVisitAlreadyExistsException("PlaceVisit "
+              + placeVisit.name() + " already exists for trip " + placeVisit.tripId()));
+        }
+        placesMap.put(placeVisit.placeId(), placeVisit);
+        return placesMap;
+      });
+    } catch (RuntimeException e) {
+      throw e.getCause();
+    }
+  }
+
+  @Override
+  public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+
+    if (placesMap.remove(placeId) == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+  }
+
+  @Override
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+  }
+
+  @Override
+  public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      return false;
```

Comment:
Now that I understand the intent of this method:  don't you want to compute a new map value if this isn't already present? (Note: you'll still want to return false somehow, which might be tricky with `computeIfAbsent`, but there are ways you can hack around that)

If that's not the intent here, then I'm still unclear on what we're doing and you'll definitely want to expand the javadoc comment to explain what's going on.